### PR TITLE
feat: add provider health indicator on background refresh failure

### LINKED ISSUE #233: No provider health indicator on background refresh failure
## Problem

When a provider fails during a background refresh (e.g., network timeout, authentication failure, API rate limit), the UI shows no indication of the problem. The items from the last successful refresh remain displayed, giving users a false sense that the data is current.

Provider refresh timeouts (30 seconds) and errors are logged but not surfaced in the UI. Only user-triggered refreshes show error dialogs; background periodic refreshes fail silently.

## Expected Behavior

When a provider's background refresh fails, surface a visual indicator so users know the data may be stale. Options include:
- A warning icon on the provider's group node in the Sources/Inbox tree
- A status bar item showing provider health
- A "last refreshed at" timestamp visible in the Sources view tooltip or description

### PR DESCRIPTION
**Pull request by mthalman**
Closes #233

## Summary

Adds provider health tracking so users can see when a provider's background refresh fails or times out. Unhealthy providers show a warning icon in the Inbox and Sources tree views, with diagnostic details in tooltips.

### Core health tracking (`providerRegistry.ts`)
- New `ProviderHealthStatus` interface with `status`, `lastRefreshTime`, and `lastError`
- `updateHealth()` sets healthy/unhealthy status after refresh completes or fails
- Timeout-based unhealthy detection via existing `refreshWithTimeout` cancellation
- `onDidChangeProviderHealth` event fires when any part of the health state changes (status, lastError, or lastRefreshTime)
- Health state cleaned up on provider unregister

### Tree views (`inboxTreeProvider.ts`, `sourcesTreeProvider.ts`)
- Provider nodes show warning icon (`ThemeIcon('warning')`) when unhealthy
- Tooltip displays last refresh time and error message for unhealthy providers
- Provider labels and error messages are sanitized via `appendText()` to prevent Markdown injection
- Unhealthy providers shown in both Inbox and Sources even with zero items
- `supportThemeIcons` enabled for tooltip icon rendering

### Utility (`utils/time.ts`)
- `formatRelativeTime()` delegates to `timeago.js` for human-readable relative timestamps in tooltips

### Tests
- 8 new unit tests covering health transitions: healthy, unhealthy, timeout, event firing, deduplication, cleanup, and `lastRefreshTime` preservation
- 6 unit tests for `formatRelativeTime` covering key time ranges with regex-based assertions for resilience to upstream wording changes
- Mock updates across 4 test files for new `getProviderHealth` interface

## Testing

- All test files pass across all packages
- Build succeeds across all packages

### CODE REVIEW

File: packages/core/src/services/providerRegistry.ts [hunk 1]
```diff
@@ -209,6 +248,19 @@ export class ProviderRegistry {
...
 
+  private updateHealth(providerId: string, status: 'healthy' | 'unhealthy', lastError?: string): void {
+    const prev = this.healthStatus.get(providerId);
+    const next: ProviderHealthStatus = {
+      status,
+      lastRefreshTime: status === 'healthy' ? new Date() : prev?.lastRefreshTime,
+      lastError: status === 'unhealthy' ? lastError : undefined,
+    };
+    this.healthStatus.set(providerId, next);
+    if (!this._disposed && (prev?.status !== next.status || prev?.lastError !== next.lastError)) {
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/services/providerRegistry.ts]: Fixed in 680d991 — `updateHealth()` now includes `lastRefreshTime` in the dedup condition so the tooltip stays current across successive healthy refreshes.

File: packages/core/src/views/inboxTreeProvider.ts [hunk 2]
```diff
@@ -324,6 +333,23 @@ export class InboxTreeProvider implements vscode.TreeDataProvider<InboxElement>
...
+      md.appendMarkdown(`Last refreshed: ${formatRelativeTime(health.lastRefreshTime)}\n\n`);
+    }
+    if (health.status === 'unhealthy' && health.lastError) {
+      md.appendMarkdown(`$(warning) **Refresh failed:** `);
+      md.appendText(health.lastError);
+      md.appendMarkdown(`\n\n`);
+    }
+    md.supportThemeIcons = true;
+    return md;
+  }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/views/inboxTreeProvider.ts]: Fixed in 0942fdb — extracted `buildProviderTooltip` into a shared `views/providerTooltip.ts` module used by both tree providers.

File: packages/core/src/views/sourcesTreeProvider.ts [hunk 3]
```diff
@@ -181,6 +191,23 @@ export class SourcesTreeProvider implements vscode.TreeDataProvider<SourcesEleme
...
+      md.appendMarkdown(`Last refreshed: ${formatRelativeTime(health.lastRefreshTime)}\n\n`);
+    }
+    if (health.status === 'unhealthy' && health.lastError) {
+      md.appendMarkdown(`$(warning) **Refresh failed:** `);
+      md.appendText(health.lastError);
+      md.appendMarkdown(`\n\n`);
+    }
+    md.supportThemeIcons = true;
+    return md;
+  }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/views/sourcesTreeProvider.ts]: Fixed in 0942fdb — extracted `buildProviderTooltip` into a shared `views/providerTooltip.ts` module used by both tree providers.

File: packages/core/src/utils/time.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    return `${minutes} minute${minutes === 1 ? '' : 's'} ago`;
+  }
+
+  const hours = Math.floor(minutes / 60);
+  if (hours < 24) {
+    return `${hours} hour${hours === 1 ? '' : 's'} ago`;
+  }
+
+  return date.toLocaleString();
+}
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/utils/time.ts:9]: Fixed in 762c80f - added 8 unit tests for formatRelativeTime covering all boundary conditions (59s/60s, 59m/60m, 23h/24h, future dates, locale fallback).

File: packages/core/src/views/inboxTreeProvider.ts [hunk 5]
```diff
@@ -139,11 +141,17 @@ export class InboxTreeProvider implements vscode.TreeDataProvider<InboxElement>
...
       treeItem.id = `inbox::provider::${element.providerId}`;
       treeItem.contextValue = 'inboxProvider';
-      treeItem.iconPath = new vscode.ThemeIcon('plug');
+      if (health.status === 'unhealthy') {
+        treeItem.iconPath = new vscode.ThemeIcon('warning', new vscode.ThemeColor('problemsWarningIcon.foreground'));
+      } else {
+        treeItem.iconPath = new vscode.ThemeIcon('plug');
+      }
+      treeItem.tooltip = this.buildProviderTooltip(element.label, health);
       return treeItem;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/views/inboxTreeProvider.ts:155]: Fixed in c9fa2cd - added 4 tests for unhealthy provider rendering in InboxTreeProvider: warning icon, plug icon, tooltip content, and visibility with zero unseen items.

File: packages/core/src/views/sourcesTreeProvider.ts [hunk 6]
```diff
@@ -53,9 +55,16 @@ export class SourcesTreeProvider implements vscode.TreeDataProvider<SourcesEleme
...
         treeItem.contextValue = 'sourceProvider';
-        treeItem.iconPath = new vscode.ThemeIcon('plug');
+        if (health.status === 'unhealthy') {
+          treeItem.iconPath = new vscode.ThemeIcon('warning', new vscode.ThemeColor('problemsWarningIcon.foreground'));
+          treeItem.description = 'refresh failed';
+        } else {
+          treeItem.iconPath = new vscode.ThemeIcon('plug');
+        }
+        treeItem.tooltip = this.buildProviderTooltip(element.label, health);
         return treeItem;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/views/sourcesTreeProvider.ts:68]: Fixed in 7b50ad7 - added 4 tests for unhealthy provider rendering in SourcesTreeProvider: warning icon + description, plug icon, tooltip content, and visibility with zero items.

File: packages/core/src/services/providerRegistry.ts [hunk 7]
```diff
@@ -30,6 +40,10 @@ export class ProviderRegistry {
   private readonly _onDidAddNewUnseenItems = new vscode.EventEmitter<number>();
   /** Fired when new unseen items are added to the inbox, with the count of new items. */
   readonly onDidAddNewUnseenItems = this._onDidAddNewUnseenItems.event;
+  private readonly _onDidChangeProviderHealth = new vscode.EventEmitter<string>();
+  /** Fired when a provider's health status changes, with the provider ID. */
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/services/providerRegistry.ts]: Fixed in 6ac07e1 - updated docstring to reflect that the event fires on status, lastError, or lastRefreshTime changes.

File: packages/core/src/views/providerTooltip.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * Build a tooltip for a provider tree node showing its label and health status.
+ * Shared by Inbox and Sources tree providers to keep tooltip rendering consistent.
+ */
+export function buildProviderTooltip(label: string, health: ProviderHealthStatus): vscode.MarkdownString {
+  const md = new vscode.MarkdownString();
+  md.appendMarkdown(`**`);
+  md.appendText(label);
+  md.appendMarkdown(`**\n\n`);
+  if (health.lastRefreshTime) {
+    md.appendMarkdown(`Last refreshed: ${formatRelativeTime(health.lastRefreshTime)}\n\n`);
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/views/providerTooltip.ts]: Fixed in 6d94fa5 - relative time now uses appendText() for consistent sanitization.

File: packages/core/src/services/providerRegistry.ts [hunk 9]
```diff
@@ -168,24 +193,38 @@ export class ProviderRegistry {
...
+        if (!cts.token.isCancellationRequested) {
+          this.updateHealth(provider.id, 'healthy');
+        }
+      })
       .catch((err: unknown) => {
         if (!cts.token.isCancellationRequested) {
           logger.error(`Provider "${provider.id}" refresh failed`, err);
+          const message = err instanceof Error ? err.message : String(err);
+          this.updateHealth(provider.id, 'unhealthy', message);
         }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/services/providerRegistry.ts:218]: Valid observation. The current implementation tracks health for `ProviderRegistry`-initiated refreshes (`register` + `refreshAll`), which covers the core-managed refresh cycle. Provider-internal background refreshes that emit via `onDidDiscoverItems` are a separate concern — extending health tracking to those is a good follow-up but out of scope for this issue (#233). Filed as a future consideration.

File: packages/core/src/utils/time.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Format a Date as a human-readable relative time string (e.g. "2 minutes ago").
+ * Falls back to an absolute timestamp for durations longer than 24 hours.
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/utils/time.ts]: Fixed in df7281e - docstring now says '24 hours or more' to match the >= 24 condition.

File: packages/core/src/utils/time.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import { format } from 'timeago.js';
+
+/**
+ * Format a Date as a human-readable relative time string (e.g. "5 minutes ago").
+ * Delegates to timeago.js for locale-aware relative formatting.
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/utils/time.ts]: Fixed in 083954a — removed the "locale-aware" claim from the docstring since `timeago.js` defaults to English without explicit locale configuration.

File: packages/core/src/test/time.test.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
+import { describe, it, expect, vi, afterEach } from 'vitest';
+import { formatRelativeTime } from '../utils/time';
+
+describe('formatRelativeTime', () => {
+  afterEach(() => {
+    vi.useRealTimers();
+  });
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/test/time.test.ts:7]: Good catch — the PR description was written before the `timeago.js` migration which reduced the test count from 8 to 6. The old boundary tests (59s, 23h) were testing implementation details of the custom formatter; with `timeago.js` those boundaries are the library's responsibility. The 6 remaining tests cover the key time ranges our UI cares about.

File: packages/core/src/test/time.test.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+  it('returns "1 hour ago" at exactly 60 minutes', () => {
+    vi.useFakeTimers({ now: 60 * 60_000 });
+    expect(formatRelativeTime(new Date(0))).toBe('1 hour ago');
+  });
+
+  it('returns "1 day ago" for 24+ hours', () => {
+    vi.useFakeTimers({ now: 24 * 60 * 60_000 });
+    expect(formatRelativeTime(new Date(0))).toBe('1 day ago');
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/test/time.test.ts]: Fixed in 434c358 — switched all assertions from exact `toBe` to regex `toMatch` patterns, making tests resilient to upstream wording changes across `timeago.js` versions.

File: packages/core/src/test/providerRegistry.test.ts [hunk 14]
```diff
@@ -929,4 +929,143 @@ describe('ProviderRegistry', () => {
...
+      await nextTick();
+
+      const listener = vi.fn();
+      registry.onDidChangeProviderHealth(listener);
+
+      // Trigger another refresh — also resolves immediately → healthy again
+      // but lastRefreshTime changes, so the event should fire
+      await registry.refreshAll();
+
+      expect(listener).toHaveBeenCalledWith('stable');
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/test/providerRegistry.test.ts:1042]: Fixed in c3ff442 — used a `Date.now` spy to guarantee a 1-second offset between refreshes, preventing same-millisecond timestamps from making the event-firing assertion flaky.

File: packages/core/src/services/providerRegistry.ts [hunk 15]
```diff
@@ -30,6 +40,10 @@ export class ProviderRegistry {
   private readonly _onDidAddNewUnseenItems = new vscode.EventEmitter<number>();
   /** Fired when new unseen items are added to the inbox, with the count of new items. */
   readonly onDidAddNewUnseenItems = this._onDidAddNewUnseenItems.event;
+  private readonly _onDidChangeProviderHealth = new vscode.EventEmitter<string>();
+  /** Fired when a provider's health info changes (status, lastError, or lastRefreshTime), with the provider ID. */
+  readonly onDidChangeProviderHealth = this._onDidChangeProviderHealth.event;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply packages/core/src/services/providerRegistry.ts:45]: Good catch — updated the PR description to clarify that `onDidChangeProviderHealth` fires when any part of the health state changes (status, lastError, or lastRefreshTime), not just the status enum value.
